Update reference documentation for Worker accessible DO interfaces (#17411)
cloudflare/cloudflare-docs: src/content/docs/durable-objects/api/id.mdx

---

> title: Durable Object ID

Reviewer: Thoughts on combining `ID` and `Stub` pages?

Author: I was thinking that we could keep them separate for consistency. It keeps the side bar navigation clean on the RHS and I'd expect the side bar navigation to be cleaned up further on the LHS as we continue down this path.

pcx_content_type: concept
sidebar:
  order: 2
---

import { Tabs, TabItem } from "~/components";

## Description

> The `DurableObjectId` interface refers to a new or existing Durable Object instance. This interface is most frequently used to by [`DurableObjectNamespace::get`](./namespace.mdx) to obtain a stub for submitting requests to the Durable Object class.

Reviewer 2: Adding fix

> Note that creating an ID for a Durable Object instance does not invoke the constructor of the Durable Object. This is done lazily when calling [`DurableObjectNamespace::get`](/durable-objects/api/namespace/#get) to create a [`DurableObjectStub`](/durable-objects/api/stub) from a `DurableObjectId`. This ensures that objects are not constructed unless their methods can be directly invoked.

Reviewer 3: ```suggestion
Note that creating an ID for a Durable Object instance does not create the Durable Object. The Durable Object is created lazily after calling [`DurableObjectNamespace::get`](/durable-objects/api/namespace/#get) to create a [`DurableObjectStub`](/durable-objects/api/stub) from a `DurableObjectId`. This ensures that objects are not constructed until they are actually accessed.
```

:::note[`DurableObjectId`]

If you are experiencing an issue with a particular Durable Object instance, you may wish to log the `DurableObjectId` from your Worker and include it in your Cloudflare support request.

:::

## Methods

### `toString`

#### Description

`toString` converts a `DurableObjectId` to a 64 digit hex string. This string is useful for logging purposes or storing the `DurableObjectId` elsewhere, for example, in a session cookie. This string can be used to reconstruct a `DurableObjectId` via `DurableObjectNamespace::idFromString`.

```js
// Create a new unique ID
const id = env.MY_DURABLE_OBJECT.newUniqueId();
// Save the unique ID elsewhere, e.g. a session cookie via id.toString()
...
// Recreate the ID from the string
const id = env.MY_DURABLE_OBJECT.idFromString(session_id);
```

#### Parameters

None.

#### Return value

A 64 digit hex string.

### `equals`

#### Description

`equals` is used to compare equality between two instances of `DurableObjectId`.

```js
const id1 = env.MY_DURABLE_OBJECT.newUniqueId();
const id2 = env.MY_DURABLE_OBJECT.newUniqueId();

> console.assert(!id1.equals(id2), "This should always be true");

Reviewer 3: ```suggestion
console.assert(!id1.equals(id2), "Different unique ids should never be equal.");
```

```

#### Parameters

A required `DurableObjectId` to compare against.

#### Return value

A boolean. True if equal and false otherwise.

## Properties

### `name`

#### Description

`name` is an optional property of a `DurableObjectId`, which returns the name that was used to create the `DurableObjectId` via [`DurableObjectNamespace::idFromName`](/durable-objects/api/namespace/#idfromname). This value is undefined if the `DurableObjectId` was constructed using [`DurableObjectNamespace::newUniqueId`](/durable-objects/api/namespace/#newuniqueid).

```js

> const id1 = env.MY_DURABLE_OBJECT.newUniqueId();

Reviewer 2: Replaced `id1` with `uniqueId`
Replaced `id2` with `fromNameId`

> const id1 = env.MY_DURABLE_OBJECT.newUniqueId();

Reviewer 2: ```suggestion
const uniqueId = env.MY_DURABLE_OBJECT.newUniqueId();
```

> const id2 = env.MY_DURABLE_OBJECT.idFromName("foo");

Reviewer 2: ```suggestion
const fromNameId = env.MY_DURABLE_OBJECT.idFromName("foo");
```

> console.assert(id1.name === undefined, "This should always be true");

Reviewer 3: ```suggestion
console.assert(id1.name === undefined, "unique ids have no name");
```

> console.assert(id1.name === undefined, "unique ids have no name");

Reviewer 2: ```suggestion
console.assert(uniqueId.name === undefined, "unique ids have no name");
```

> console.assert(id2.name === "foo", "This should always be true");

Reviewer 3: ```suggestion
console.assert(id2.name === "foo", "name matches parameter to idFromName");
```

> console.assert(id2.name === "foo", "name matches parameter to idFromName");

Reviewer 2: ```suggestion
console.assert(fromNameId.name === "foo", "name matches parameter to idFromName");
```
